rfd/0144-client-tools-updates.md

```diff
@@ -0,0 +1,295 @@
+---
+authors: Russell Jones (rjones@goteleport.com) and Bernard Kim (bernard@goteleport.com)
+state: draft
+---
+
+# RFD 0144 - Client Tools Updates
+
+## Required Approvers
+
+* Engineering: @sclevine && @bernardjkim && @r0mant
+* Product: @klizhentas || @xinding33
+* Security: @reedloden
+
+## What/Why
+
+This RFD describes how client tools like `tsh` and `tctl` can be kept up to
+date, either using automatic updates or self-managed updates.
+
+Keeping client tools updated helps with security (fixes for known security
+vulnerabilities are pushed to endpoints), bugs (fixes for resolved issues are
+pushed to endpoints), and compatibility (users no longer have to learn and
+understand [Teleport component
+compatibility](https://goteleport.com/docs/upgrading/overview/#component-compatibility)
+rules).
+
+## Details
+
+### Summary
+
+Client tools like `tsh` and `tctl` will automatically download and install the
+required version for the Teleport cluster.
+
+Enrollment in automatic updates for client tools will be controlled at the
+cluster level. By default all clusters will be opted into automatic updates for
+client tools. Cluster administrators using MDM software like JamF will be able
+opt-out manually manage updates.
+
+Inspiration drawn from https://go.dev/doc/toolchain.
+
+### Implementation
+
+#### Client tools
+
+##### Automatic updates
+
+When `tsh login` is executed, client tools will check `/v1/webapi/ping` to
+determine if automatic updates are enabled. If the cluster's required version
+differs from the current binary, client tools will download and re-execute
+using the version required by the cluster.
+
+The original client tools binaries won't be overwritten. Instead, additional
+binaries will be downloaded and stored in `~/.tsh/bin` with `0555` permissions.
```

Comment:
Note that this makes `.tsh` slightly more annoying to delete:
```
teleport:cloud stephen$ mkdir .tempdir
teleport:cloud stephen$ touch .tempdir/file
teleport:cloud stephen$ chmod 555 .tempdir/file
teleport:cloud stephen$ rm -r .tempdir/
override r-xr-xr-x stephen/staff for .tempdir//file? y
```
Probably okay, but `-w` on the owner doesn't provide much security benefit.

Might suggest `~umask&0755`